File: android/app/src/main/java/com/nextpage/data/remote/addons/AddonCatalogProvider.kt
package com.nextpage.data.remote.addons

import com.nextpage.data.remote.catalog.CatalogBook
import com.nextpage.data.remote.catalog.CatalogErrorCode
import com.nextpage.data.remote.catalog.CatalogProvider
import com.nextpage.data.remote.catalog.CatalogSourceInfo
import com.nextpage.data.remote.catalog.CatalogSourceKind
import com.nextpage.data.remote.catalog.MAX_PAGE_SIZE
import com.nextpage.data.remote.catalog.PagedResult
import com.nextpage.data.remote.catalog.addonSource
import com.nextpage.data.remote.catalog.catalogError
import com.nextpage.data.remote.catalog.computeNextPage
import com.nextpage.data.remote.catalog.isHttpSuccess
import com.nextpage.data.remote.catalog.RETRY_BASE_DELAY_MS
import com.nextpage.data.remote.catalog.shouldRetryStatus
import kotlinx.coroutines.delay
import com.nextpage.data.remote.catalog.mapHttpStatusToCode
import com.nextpage.data.remote.catalog.resolveDownloadUrl
import com.nextpage.domain.access.AccessGroup
import com.nextpage.domain.access.AccessOption
import com.nextpage.domain.access.LegalAccess
import com.nextpage.domain.access.resolveAccess as resolveU3Access
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * U4 external-open link title (localized copy, if any, belongs to U5).
 */
private const val TITLE_OPEN_EXTERNAL = "Open"

/**
 * U4 v2 resolve item: per-item `accessType`/`license`/`readUrl`/
 * `downloadUrl`. Unknown access types parse to null and fail closed
 * ([mayDownloadInApp] requires an explicit free or cleared license, so an
 * unknown item is never downloadable); its `readUrl` may still surface as
 * an external-open link.
 */
data class AddonResolveItem(
    val accessType: AddonAccessType?,
    val license: String?,
    val readUrl: String?,
    val downloadUrl: String?
) {
    /** Only `free` (or license-cleared) items with an https download may flow in-app. */
    val mayDownloadInApp: Boolean
        get() = (accessType == AddonAccessType.FREE || ManifestValidator.isLicenseCleared(license)) &&
            downloadUrl != null

    /** External-open link for this item, or null when it has no https read URL. */
    fun toExternalOption(): AccessOption? {
        val url = readUrl ?: return null
        val group = when (accessType) {
            AddonAccessType.BUY -> AccessGroup.BUY
            AddonAccessType.SUBSCRIBE -> AccessGroup.SUBSCRIBE
            else -> AccessGroup.FREE
        }
        return AccessOption(group, TITLE_OPEN_EXTERNAL, url, opensInApp = false)
    }
}

/**
 * One [CatalogProvider] per installed addon manifest. Sources are
 * `addon:<addonId>`. Manifests MAY declare endpoint templates
 * (searchUrl/detailsUrl, validated https by [ManifestValidator]); when
 * present, search/getDetails fetch the addon's own catalog payloads through
 * the platform network layer (KtorAddonHttpTransport over the shared catalog
 * HttpClient) and parse them into CatalogBooks with provider = the addon
 * source id. Endpoint-less manifests stay browse-only: stable empty page /
 * NOT_FOUND, zero I/O. Mirrors desktop AddonCatalogProvider.ts.
 *
 * U4: v2 manifests MAY also declare `resolveUrl` + `capabilities`.
 * [resolveAccess] renders `resolveUrl` from the book identity, but only
 * after per-addon disclosure consent was recorded in [consent] — without
 * consent it returns an empty [LegalAccess] with zero I/O.
 */
class AddonCatalogProvider(
    manifest: AddonManifest,
    addonId: String,
    private val transport: AddonHttpTransport?,
    private val retryDelayMs: Long = RETRY_BASE_DELAY_MS,
    private val consent: AddonConsentStore = InMemoryAddonConsentStore()
) : CatalogProvider {

    init {
        val hasEndpoint = manifest.searchUrl != null || manifest.detailsUrl != null
        require(!(hasEndpoint && transport == null)) {
            "addon declares catalog endpoints but no transport is configured"
        }
    }

    private val source = CatalogSourceInfo(
        sourceId = addonSource(addonId),
        name = manifest.name,
        kind = CatalogSourceKind.ADDON
    )

    private val searchUrl: String? = manifest.searchUrl
    private val detailsUrl: String? = manifest.detailsUrl
    private val resolveUrl: String? = manifest.resolveUrl
    private val addonIdKey: String = addonId

    /** U4: declared v2 capabilities, disclosed before consent is recorded (U5 UI). */
    val capabilities: List<String> = manifest.capabilities

    private val json = Json { ignoreUnknownKeys = true }

    override fun listSources(): List<CatalogSourceInfo> = listOf(source)

    override suspend fun search(query: String, page: Int): PagedResult {
        val template = searchUrl ?: return BROWSE_ONLY_PAGE
        val url = renderTemplate(
            template,
            mapOf("query" to encodeValue(query), "page" to page.toString())
        )
        return parseSearchPayload(fetchJson(url), source.sourceId, page)
    }

    override suspend fun getDetails(id: String): CatalogBook {
        val prefix = "${source.sourceId}:"
        if (!id.startsWith(prefix) || id.length <= prefix.length) {
            throw catalogError(CatalogErrorCode.NOT_FOUND, "unknown catalog id $id")
        }
        val template = detailsUrl
            ?: throw catalogError(CatalogErrorCode.NOT_FOUND, "addon source has no details endpoint: $id")
        val url = renderTemplate(
            template,
            mapOf("bookId" to encodeValue(id.removePrefix(prefix)))
        )
        return parseAddonBook(fetchJson(url), id, source.sourceId)
    }

    private suspend fun fetchJson(url: String): JsonElement {
        var response = callTransport(url)
        if (shouldRetryStatus(response.status)) {
            if (retryDelayMs > 0) delay(retryDelayMs)
            response = callTransport(url)
        }
        if (!response.status.isHttpSuccess()) {
            throw catalogError(mapHttpStatusToCode(response.status), "addon payload status ${response.status}")
        }
        if (response.body.size > ManifestValidator.MAX_MANIFEST_BYTES) {
            throw catalogError(CatalogErrorCode.UPSTREAM_ERROR, "addon payload exceeds size cap")
        }
        return try {
            json.parseToJsonElement(response.body.decodeToString())
        } catch (_: Exception) {
            throw catalogError(CatalogErrorCode.UPSTREAM_ERROR, "addon payload is not valid JSON")
        }
    }

    private suspend fun callTransport(url: String): AddonResource {
        val transport = transport
            ?: throw catalogError(CatalogErrorCode.NETWORK_ERROR, "addon transport unavailable")
        return try {
            transport.fetch(url)
        } catch (err: Throwable) {
            if (err is com.nextpage.data.remote.catalog.CatalogException) throw err
            throw catalogError(CatalogErrorCode.NETWORK_ERROR, "addon payload request failed")
        }
    }

    override fun resolveDownloadUrl(formats: Map<String, String>, preferEpub: Boolean): String =
        resolveDownloadUrl(formats, preferEpub)

    /**
     * U4 consent-gated resolve. Without recorded consent for this addon —
     * or without a `resolveUrl` capability — returns an empty [LegalAccess]
     * with zero I/O. With consent, `resolveUrl` renders from the book
     * identity (`{isbn}`, `{title}`, `{author}`, `{openLibraryId}`,
     * `{googleBooksId}`), the free-gate admits only `free`/license-cleared
     * items to the in-app download path, everything else surfaces as
     * external-open links, and the in-app download additionally reuses the
     * U3 gate (`isPublicDomain == true` on the book).
     */
    suspend fun resolveAccess(book: CatalogBook): LegalAccess {
        if (!consent.hasConsent(addonIdKey)) return LegalAccess(book.id, false, null, emptyList())
        val template = resolveUrl ?: return LegalAccess(book.id, false, null, emptyList())
        val items = parseResolvePayload(fetchJson(renderTemplate(template, resolveParams(book))))
        val candidate = items.firstOrNull { it.mayDownloadInApp }?.downloadUrl
        val gate = resolveU3Access(book.copy(downloadUrl = candidate))
        return LegalAccess(
            bookId = book.id,
            canDownloadInApp = gate.canDownloadInApp,
            downloadUrl = gate.downloadUrl,
            options = items.mapNotNull { it.toExternalOption() }
        )
    }

    companion object {
        /** Parity with desktop encodeURIComponent: spaces are %20, never +. */
        private fun encodeValue(value: String): String =
            java.net.URLEncoder.encode(value, "UTF-8").replace("+", "%20")

        private val BROWSE_ONLY_PAGE = PagedResult(emptyList(), null, 0)

        /** U4 resolveUrl template keys, rendered from the U3 book identity. */
        private const val PARAM_ISBN = "isbn"
        private const val PARAM_TITLE = "title"
        private const val PARAM_AUTHOR = "author"
        private const val PARAM_OPEN_LIBRARY_ID = "openLibraryId"
        private const val PARAM_GOOGLE_BOOKS_ID = "googleBooksId"

        /** Substitute `{placeholders}` in a validated endpoint template. */
        internal fun renderTemplate(template: String, params: Map<String, String>): String =
            Regex("\\{(\\w+)\\}").replace(template) { match ->
                params[match.groupValues[1]] ?: match.value
            }

        /** U4: identity params for a resolveUrl template (missing identity ⇒ empty). */
        private fun resolveParams(book: CatalogBook): Map<String, String> = mapOf(
            PARAM_ISBN to encodeValue(book.isbn13 ?: book.isbn10 ?: ""),
            PARAM_TITLE to encodeValue(book.title),
            PARAM_AUTHOR to encodeValue(book.authors.firstOrNull() ?: ""),
            PARAM_OPEN_LIBRARY_ID to encodeValue(book.openLibraryWorkId ?: ""),
            PARAM_GOOGLE_BOOKS_ID to encodeValue(book.googleBooksId ?: "")
        )

        private fun httpsOrNull(value: JsonElement?): String? =
            ((value as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content)
                ?.takeIf { it.startsWith("https://") }

        private fun stringArray(value: JsonElement?, field: String): List<String> {
            if (value == null || value is JsonNull) return emptyList()
            val array = value as? JsonArray
                ?: throw catalogError(CatalogErrorCode.UPSTREAM_ERROR, "malformed addon payload: $field must be an array of strings")
            return array.map { element ->
                (element as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content
                    ?: throw catalogError(CatalogErrorCode.UPSTREAM_ERROR, "malformed addon payload: $field must be strings")
            }
        }

        private fun malformed(detail: String): Nothing =
            throw catalogError(CatalogErrorCode.UPSTREAM_ERROR, "malformed addon payload: $detail")

        private fun parseAddonBook(value: JsonElement, bookId: String, sourceId: String): CatalogBook {
            val obj = value as? JsonObject ?: malformed("book must be an object")
            val rawId = (obj["id"] as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content
            val rawTitle = (obj["title"] as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content
            if (rawId.isNullOrEmpty() || rawTitle.isNullOrEmpty()) malformed("book missing id/title")
            return CatalogBook(
                id = bookId,
                provider = sourceId,
                title = rawTitle,
                authors = stringArray(obj["authors"], "authors"),
                coverUrl = httpsOrNull(obj["coverUrl"]),
                languages = stringArray(obj["languages"], "languages"),
                subjects = stringArray(obj["subjects"], "subjects"),
                downloadUrl = httpsOrNull(obj["downloadUrl"])
            )
        }

        /**
         * U4: parse a resolve payload (`{"results": [...]}`) into access
         * items. `accessType`/`license`/`readUrl`/`downloadUrl` are all
         * optional per item; non-https URLs are dropped (never fail the
         * whole payload); unknown `accessType` parses to null (fail closed
         * for the download path at [AddonResolveItem.mayDownloadInApp]).
         */
        internal fun parseResolvePayload(payload: JsonElement): List<AddonResolveItem> {
            val obj = payload as? JsonObject ?: malformed("resolve payload must be an object with a results array")
            val results = obj["results"] as? JsonArray ?: malformed("resolve payload must be an object with a results array")
            return results.mapIndexed { i, element ->
                val entry = element as? JsonObject ?: malformed("resolve item $i must be an object")
                AddonResolveItem(
                    accessType = ManifestValidator.parseAccessType(
                        (entry["accessType"] as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content
                    ),
                    license = (entry["license"] as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content,
                    readUrl = httpsOrNull(entry["readUrl"]),
                    downloadUrl = httpsOrNull(entry["downloadUrl"])
                )
            }
        }

        internal fun parseSearchPayload(payload: JsonElement, sourceId: String, page: Int): PagedResult {
            val obj = payload as? JsonObject ?: malformed("payload must be an object with a results array")
            val results = obj["results"] as? JsonArray ?: malformed("payload must be an object with a results array")
            var totalCount = results.size
            val declared = obj["totalCount"]
            if (declared != null && declared !is JsonNull) {
                val n = (declared as? JsonPrimitive)?.content?.toIntOrNull()
                    ?: malformed("totalCount must be a non-negative integer")
                if (n < 0) malformed("totalCount must be a non-negative integer")
                totalCount = n
            }
            val books = results.mapIndexed { i, element ->
                val entry = element as? JsonObject ?: malformed("book $i must be an object")
                val rawId = (entry["id"] as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content
                if (rawId.isNullOrEmpty()) malformed("book $i missing id")
                parseAddonBook(entry, "$sourceId:${rawId}", sourceId)
            }
            val clamped = books.take(MAX_PAGE_SIZE)
            return PagedResult(
                results = clamped,
                nextPage = computeNextPage(page, clamped.size, totalCount),
                totalCount = totalCount
            )
        }
    }
}

File: desktop/src/lib/shared/services/addons/AddonCatalogProvider.ts
/**
 * AddonCatalogProvider — one CatalogProvider per installed addon manifest.
 * Sources are `addon:<addonId>`. Manifests MAY declare endpoint templates
 * (searchUrl/detailsUrl, validated https by the manifest validator); when
 * present, search/getDetails fetch the addon's own catalog payloads through
 * the platform network layer (desktop: the fetchAddonResource Rust command)
 * and parse them into CatalogBooks with provider = the addon source id.
 * Endpoint-less manifests stay browse-only: stable empty page / NOT_FOUND,
 * zero I/O. Malformed payloads reject the whole page with a stable error.
 */
import { catalogError, mapHttpStatusToCode } from '../catalog/errors';
import { backoffDelayMs, MAX_PAGE_SIZE, shouldRetryStatus } from '../catalog/policy';
import type { AddonConsentGate } from './AddonConsent';
import type {
  CatalogBook,
  CatalogFeaturedSort,
  CatalogProvider,
  CatalogSource,
  CatalogSourceInfo,
  PagedResult,
} from '../catalog/CatalogProvider';
import type { AddonAccessResolution } from '../catalog/CatalogProvider';
import type { AccessGroup, AccessOption } from '../catalog/accessResolver';
import { addonSource } from '../catalog/CatalogProvider';
import { computeNextPage, resolveDownloadUrl } from '../catalog/mappers';
import {
  MAX_MANIFEST_BYTES,
  declaredCapabilities,
  type AddonManifest,
} from '@nextpage/manifest-validator';
import { defaultAddonTransport, type AddonTransport } from './AddonRegistry';
import type { AddonFetchResult } from './AddonRegistry';

const EMPTY_PAGE: PagedResult = { results: [], nextPage: null, totalCount: 0 };

function isNonEmptyString(value: unknown): value is string {
  return typeof value === 'string' && value.length > 0;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function malformed(detail: string): never {
  throw catalogError('UPSTREAM_ERROR', `malformed addon payload: ${detail}`);
}

/** Substitute `{placeholders}` in a validated endpoint template. */
function renderTemplate(template: string, params: Record<string, string>): string {
  return template.replace(/\{(\w+)\}/g, (match, key: string) => params[key] ?? match);
}

/** Optional cover/download URLs: https strings pass, anything else maps to null. */
function httpsOrNull(value: unknown): string | null {
  return typeof value === 'string' && value.startsWith('https://') ? value : null;
}

/** Optional string-array field: absent → []; anything else → stable malformed error. */
function stringArray(value: unknown, field: string): string[] {
  if (value === undefined) return [];
  if (!Array.isArray(value) || !value.every((v) => typeof v === 'string')) {
    malformed(`${field} must be an array of strings`);
  }
  return value as string[];
}

function parseAddonBook(value: unknown, bookId: string, sourceId: string): CatalogBook {
  if (!isPlainObject(value)) malformed('book must be an object');
  if (!isNonEmptyString(value.id) || !isNonEmptyString(value.title)) {
    malformed('book missing id/title');
  }
  return {
    id: bookId,
    provider: sourceId,
    title: value.title,
    authors: stringArray(value.authors, 'authors'),
    coverUrl: httpsOrNull(value.coverUrl),
    languages: stringArray(value.languages, 'languages'),
    subjects: stringArray(value.subjects, 'subjects'),
    downloadUrl: httpsOrNull(value.downloadUrl),
  };
}

function parseSearchPayload(payload: unknown, sourceId: string, page: number): PagedResult {
  if (!isPlainObject(payload) || !Array.isArray(payload.results)) {
    malformed('payload must be an object with a results array');
  }
  let totalCount = payload.results.length;
  if (payload.totalCount !== undefined) {
    if (
      typeof payload.totalCount !== 'number' ||
      !Number.isInteger(payload.totalCount) ||
      payload.totalCount < 0
    ) {
      malformed('totalCount must be a non-negative integer');
    }
    totalCount = payload.totalCount;
  }
  const books = payload.results.map((entry, i) => {
    if (!isPlainObject(entry) || !isNonEmptyString(entry.id)) malformed(`book ${i} missing id`);
    return parseAddonBook(entry, `${sourceId}:${entry.id as string}`, sourceId);
  });
  const clamped = books.slice(0, MAX_PAGE_SIZE);
  return {
    results: clamped,
    nextPage: computeNextPage(page, clamped.length, totalCount),
    totalCount,
  };
}

/**
 * Fail-closed default gate: a provider built without an explicit consent gate
 * (tests, one-off constructions) denies every resolve. Production providers
 * always receive the shared `addonConsent` singleton via `liveComposite`.
 */
const denyAllAddonConsent: AddonConsentGate = {
  ensureLoaded: () => Promise.resolve(),
  hasConsent: () => false,
};

/**
 * Addon access resolution (slice 9): the consent gate, the no-endpoint and
 * no-capability empty results (all zero I/O), then the `resolveUrl` fetch
 * with access-item parsing, the in-app download gate, and the external
 * options. Re-exports the port types so resolution consumers import one
 * module.
 */
export type { AddonAccessResolution };
export type { AccessOption };

/** The only known v2 capability id: the addon may resolve reading access. */
export const RESOLVE_CAPABILITY = 'resolve';

/** Per-item `accessType` wire values (open set on the wire; unknown ⇒ null). */
export type AddonAccessType = 'free' | 'buy' | 'subscribe';

/**
 * Parse a v2 per-item `accessType` wire value (Android
 * `ManifestValidator.parseAccessType` parity: trimmed, case-insensitive).
 * Unknown or missing values return null so resolve callers fail closed
 * (never free).
 */
export function parseAccessType(raw: unknown): AddonAccessType | null {
  if (typeof raw !== 'string') return null;
  switch (raw.trim().toLowerCase()) {
    case 'free':
      return 'free';
    case 'buy':
      return 'buy';
    case 'subscribe':
      return 'subscribe';
    default:
      return null;
  }
}

/** License tokens cleared for the in-app download path (closed set, Android parity). */
const LICENSE_CLEARED_TOKENS: ReadonlySet<string> = new Set([
  'public-domain',
  'public domain',
  'cc0',
  'cc0-1.0',
  'pd',
]);

/** True only for license tokens in the closed cleared set (case-insensitive). */
export function isLicenseCleared(license: string | null): boolean {
  return license !== null && LICENSE_CLEARED_TOKENS.has(license.trim().toLowerCase());
}

/**
 * One parsed resolve item. `readUrl`/`downloadUrl` are https-only (anything
 * else parsed to null at the boundary); `accessType` is null for unknown
 * wire values (fail closed for the download path at `mayDownloadInApp`).
 */
export interface AddonResolveItem {
  accessType: AddonAccessType | null;
  license: string | null;
  readUrl: string | null;
  downloadUrl: string | null;
  /** Only `free` (or license-cleared) items with an https download may flow in-app. */
  mayDownloadInApp: boolean;
}

function accessGroupOf(accessType: AddonAccessType | null): AccessGroup {
  switch (accessType) {
    case 'buy':
      return 'BUY';
    case 'subscribe':
      return 'SUBSCRIBE';
    default:
      return 'FREE';
  }
}

/** Identity params for a `resolveUrl` template (missing identity ⇒ empty). */
function resolveParams(book: CatalogBook): Record<string, string> {
  return {
    isbn: encodeURIComponent(book.isbn13 ?? book.isbn10 ?? ''),
    title: encodeURIComponent(book.title),
    author: encodeURIComponent(book.authors[0] ?? ''),
    openLibraryId: encodeURIComponent(book.openLibraryWorkId ?? ''),
    googleBooksId: encodeURIComponent(book.googleBooksId ?? ''),
  };
}

/**
 * Parse a resolve payload (`{results: [...]}`) into access items
 * (Android `parseResolvePayload` parity). `accessType`/`license`/`readUrl`/
 * `downloadUrl` are all optional per item; non-https URLs are dropped per
 * item (never fail the whole payload); unknown `accessType` parses to null.
 */
function parseResolvePayload(payload: unknown): AddonResolveItem[] {
  if (!isPlainObject(payload) || !Array.isArray(payload.results)) {
    malformed('resolve payload must be an object with a results array');
  }
  return payload.results.map((entry, i) => {
    if (!isPlainObject(entry)) malformed(`resolve item ${i} must be an object`);
    const accessType = parseAccessType(entry.accessType);
    const license = typeof entry.license === 'string' ? entry.license : null;
    const readUrl = httpsOrNull(entry.readUrl);
    const downloadUrl = httpsOrNull(entry.downloadUrl);
    return {
      accessType,
      license,
      readUrl,
      downloadUrl,
      mayDownloadInApp:
        (accessType === 'free' || isLicenseCleared(license)) && downloadUrl !== null,
    } satisfies AddonResolveItem;
  });
}

/** Empty resolution: no in-app download, no external options, zero I/O. */
export const EMPTY_ADDON_ACCESS: AddonAccessResolution = {
  canDownloadInApp: false,
  downloadUrl: null,
  options: [],
};

export class AddonCatalogProvider implements CatalogProvider {
  private readonly source: CatalogSourceInfo;
  private readonly addonId: string;
  private readonly manifest: AddonManifest;
  private readonly searchUrl?: string;
  private readonly detailsUrl?: string;
  private readonly consent: AddonConsentGate;

  constructor(
    manifest: AddonManifest,
    addonId: string,
    private readonly transport: AddonTransport = defaultAddonTransport(),
    private readonly retryDelayMs: number = backoffDelayMs(0),
    consent: AddonConsentGate = denyAllAddonConsent,
  ) {
    this.source = {
      sourceId: addonSource(addonId),
      name: manifest.name,
      kind: 'addon' as const,
    };
    this.addonId = addonId;
    this.manifest = manifest;
    this.searchUrl = manifest.searchUrl;
    this.detailsUrl = manifest.detailsUrl;
    this.consent = consent;
  }

  listSources(): CatalogSourceInfo[] {
    return [this.source];
  }

  async search(query: string, page: number): Promise<PagedResult> {
    if (!this.searchUrl) return EMPTY_PAGE;
    const url = renderTemplate(this.searchUrl, {
      query: encodeURIComponent(query),
      page: String(page),
    });
    const payload = await this.fetchJson(url);
    return parseSearchPayload(payload, this.source.sourceId, page);
  }

  async getDetails(id: string): Promise<CatalogBook> {
    const prefix = `${this.source.sourceId}:`;
    if (!id.startsWith(prefix) || id.length <= prefix.length) {
      throw catalogError('NOT_FOUND', `unknown catalog id ${id}`);
    }
    if (!this.detailsUrl) {
      throw catalogError('NOT_FOUND', `addon source has no details endpoint: ${id}`);
    }
    const url = renderTemplate(this.detailsUrl, {
      bookId: encodeURIComponent(id.slice(prefix.length)),
    });
    return parseAddonBook(await this.fetchJson(url), id, this.source.sourceId);
  }

  /** Addons opt out of featured rails; the rail auto-hides (fail-closed). */
  async featured(_sort: CatalogFeaturedSort, limit: number): Promise<PagedResult> {
    if (!Number.isInteger(limit) || limit < 1) {
      throw catalogError('INVALID_PAGE', `limit must be >= 1, got ${limit}`);
    }
    return EMPTY_PAGE;
  }

  supportsFeatured(_sort: CatalogFeaturedSort): boolean {
    return false;
  }

  /**
   * Per-source search scoped to this addon's own source; anything else fails
   * closed with an empty page (never a crash, never a silent composite search).
   */
  async searchSource(sourceId: CatalogSource, query: string, page: number): Promise<PagedResult> {
    if (sourceId !== this.source.sourceId) {
      return EMPTY_PAGE;
    }
    return this.search(query, page);
  }

  /**
   * Consent-gated resolve (slice 9). Consent gates RESOLVE ONLY:
   * `search` / `getDetails` stay ungated (Android parity + the spec's
   * "resolve operation" phrasing).
   *
   * - No consent ⇒ rejects `CONSENT_REQUIRED` with ZERO transport calls.
   * - No `resolveUrl` (every v1 manifest) ⇒ empty resolution, zero I/O.
   * - Declared `capabilities` without `resolve` ⇒ same empty result, zero
   *   I/O. An UNDECLARED list leaves `resolveUrl` authoritative, preserving
   *   v1-style manifests.
   * - Otherwise the `resolveUrl` renders from the book identity and fetches
   *   through the existing transport (retry + 64 KiB cap + JSON parse).
   */
  async resolveAddonAccess(book: CatalogBook): Promise<AddonAccessResolution> {
    await this.consent.ensureLoaded();
    if (!this.consent.hasConsent(this.addonId)) {
      throw catalogError(
        'CONSENT_REQUIRED',
        `addon has no network consent: ${this.source.sourceId}`,
      );
    }
    const resolveUrl = this.manifest.resolveUrl;
    if (typeof resolveUrl !== 'string' || resolveUrl.length === 0) {
      return { ...EMPTY_ADDON_ACCESS, options: [] };
    }
    if (
      this.manifest.capabilities !== undefined &&
      !declaredCapabilities(this.manifest).includes(RESOLVE_CAPABILITY)
    ) {
      return { ...EMPTY_ADDON_ACCESS, options: [] };
    }
    const items = parseResolvePayload(
      await this.fetchJson(renderTemplate(resolveUrl, resolveParams(book))),
    );
    const candidate = items.find((item) => item.mayDownloadInApp)?.downloadUrl ?? null;
    const canDownloadInApp = candidate !== null && book.isPublicDomain === true;
    return {
      canDownloadInApp,
      downloadUrl: canDownloadInApp ? candidate : null,
      options: items.flatMap((item) =>
        item.readUrl === null
          ? []
          : [
              {
                group: accessGroupOf(item.accessType),
                titleKey: 'discover.accessOpen',
                url: item.readUrl,
                opensInApp: false,
              } satisfies AccessOption,
            ],
      ),
    };
  }

  private async fetchJson(url: string): Promise<unknown> {
    let res = await this.callTransport(url);
    if (shouldRetryStatus(res.status)) {
      await new Promise((resolve) => setTimeout(resolve, this.retryDelayMs));
      res = await this.callTransport(url);
    }
    if (res.status < 200 || res.status >= 300) {
      throw catalogError(mapHttpStatusToCode(res.status), `addon payload status ${res.status}`);
    }
    if (res.body.byteLength > MAX_MANIFEST_BYTES) {
      throw catalogError('UPSTREAM_ERROR', 'addon payload exceeds size cap');
    }
    try {
      return JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(res.body));
    } catch {
      throw catalogError('UPSTREAM_ERROR', 'addon payload is not valid JSON');
    }
  }

  private async callTransport(url: string): Promise<AddonFetchResult> {
    try {
      return await this.transport(url);
    } catch {
      throw catalogError('NETWORK_ERROR', 'addon payload request failed');
    }
  }

  resolveDownloadUrl(formats: Record<string, string>, preferEpub: boolean): string {
    return resolveDownloadUrl(formats, preferEpub);
  }
}
